### PR TITLE
Read config.json on app start

### LINKED ISSUE #609: Pickup config.json changes at app startup rather than build
Currently, the config.json file is only read at build time. To change for example the default home server, it would be nice if people deploying Hydrogen wouldn't have to run the build again.

One option would be to just disable caching for config.json like we do for index.html and read that before bootstrapping the app, but this would break offline and also cause a delay before the UI can be shown.

So we also need a caching strategy in the service worker that always returns what is in cache for config.json, if anything, and in the background updates it from the server if possible.

If caching is correctly disabled on the server, then applying config changes should just be a refresh away.

### PR DESCRIPTION
**Pull request by MidhunSureshR**
For #609;
- Platform ctor optionally accepts a config URL which is read in `init()`
-  Has vite-config changes to produce config.json during build
-  Special caching strategy for config.json

### CODE REVIEW

File: src/platform/web/Platform.js [hunk 1]
```diff
@@ -142,7 +143,7 @@ export class Platform {
             this._serviceWorkerHandler = new ServiceWorkerHandler();
             this._serviceWorkerHandler.registerAndStart(assetPaths.serviceWorker);
         }
-        this.notificationService = new NotificationService(this._serviceWorkerHandler, config.push);
+        this.notificationService = null;
```
Comment (bwindels) [inline src/platform/web/Platform.js]: can we use `undefined`? It's easier to type once we switch to typescript.

File: src/platform/web/Platform.js [hunk 2]
```diff
@@ -126,10 +126,11 @@ function adaptUIOnVisualViewportResize(container) {
 }
 
 export class Platform {
-    constructor(container, assetPaths, config, options = null, cryptoExtras = null) {
+    constructor({ container, assetPaths, config, configURL, options = null, cryptoExtras = null }) {
```
Comment (bwindels) [inline src/platform/web/Platform.js:129]: You'll need to adjust the LegacyPlatform ctor as well

File: src/platform/web/sw.js [hunk 3]
```diff
@@ -119,6 +126,32 @@ async function handleRequest(request) {
     }
 }
 
+async function handleConfigRequest(request) {
+    const url = new URL(request.url);
+    // rewrite / to /index.html so it hits the cache
+    if (url.origin === baseURL.origin && url.pathname === baseURL.pathname) {
```
Comment (bwindels) [inline src/platform/web/sw.js]: I don't think you need to rewrite / to index.html for the config?

File: src/platform/web/sw.js [hunk 4]
```diff
@@ -119,6 +126,32 @@ async function handleRequest(request) {
...
+        request = new Request(new URL("index.html", baseURL.href));
+    }
+    let response = await readCache(request);
+    if (response) {
+        fetchAndUpdateConfig(request);
```
Comment (bwindels) [inline src/platform/web/sw.js]: perhaps you could unify the code paths to make it a bit easier to read:
```
const networkResponsePromise = fetchAdnUpdateConfig(request);
if (response) {
  return response;
} else {
  return await networkResponsePromise;
}
```

File: vite.config.js [hunk 5]
```diff
@@ -14,6 +14,11 @@ export default defineConfig(({mode}) => {
             outDir: "../../../target",
             minify: true,
             sourcemap: true,
+            rollupOptions: {
+                output: {
+                    assetFileNames: (asset) => asset.name.includes("config.json") ? "assets/[name][extname]": "assets/[name].[hash][extname]",
```
Comment (bwindels) [inline vite.config.js:19]: we use this option elsewhere as well to not hash the theme manifests, right? do we have to take that into account here or not?

**Review by bwindels — APPROVED** [review]
looks good, just a few nits. Did you test the app still works offline?

File: src/platform/web/sw.js [hunk 6]
```diff
@@ -75,7 +75,11 @@ self.addEventListener('fetch', (event) => {
     This has to do with xhr not being supported in service workers.
     */
     if (event.request.method === "GET") {
-        event.respondWith(handleRequest(event.request));
+        if (event.request.url.includes("config.json")) {
```
Comment (bwindels) [inline src/platform/web/sw.js]: actually, can we check for config.json in handleRequest and call handleConfigRequest from there?

File: vite.config.js (hunk 5 above)
Comment (MidhunSureshR) [reply vite.config.js:19]: I haven't created that PR yet. Those changes will be in the theme chooser PR.

**Comment by MidhunSureshR** [comment]
> Did you test the app still works offline?

Yep :+1:

File: src/platform/web/sw.js [hunk 7]
```diff
@@ -96,7 +96,7 @@ let pendingFetchAbortController = new AbortController();
 async function handleRequest(request) {
     try {
         if (request.url.includes("config.json")) {
-            return handleConfigRequest(request);
+            return await handleConfigRequest(request);
```
Comment (bwindels) [inline src/platform/web/sw.js]: you actually don't need this I think? :shrug:

**Review by bwindels — APPROVED** [review]
looks great, this looks good to go!

**Pull request merged**